core/src/main/java/org/jdbi/v3/core/ResultProducer.java

```diff
@@ -13,25 +13,24 @@
  */
 package org.jdbi.v3.core;
 
+import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 
-import org.jdbi.v3.core.mapper.RowMapper;
-
 /**
- * A StatementExecutor generates the mapped results of a {@link SqlStatement},
+ * A {@code ResultProducer} generates the mapped results of a {@link SqlStatement},
  * potentially wrapped in a container object.
  */
 @FunctionalInterface
-public interface StatementExecutor<T, R>
+public interface ResultProducer<T, R>
 {
     /**
      * Produce a statement result.
-     * @param mapper the mapper the user's code told you to use
-     * @param rs the result set
+     * @param stmt the prepared statement
+     * @param rs the result set to iterate
      * @param ctx the statement context
      * @return an object of the type your caller expects
      * @throws SQLException sadness
      */
-    R execute(RowMapper<T> mapper, ResultSet rs, StatementContext ctx) throws SQLException;
+    R produce(PreparedStatement stmt, ResultSet rs, StatementContext ctx) throws SQLException;
```

Comment:
I was also thinking about making `ResultIterable` its own type, and rather than a subtype of `ResultBearing`. That way result producers like e.g. `returningMappedResults` could return a  `ResultIterable`.

Thus:

```java
stmt.execute(returningMappedResults(Contact.class)).findFirst()
stmt.execute(returningMappedResults(Contact.class)).list()
stmt.execute(returningMappedResults(Contact.class)).collect(toSet())
```

Also here's how I could see multiple return result sets working with this:

```
stmt.execute(returningMultipleResultSets(Contact.class, Account.class, Invoice.class))
    .listOfLists();

stmt.execute(
  returningMappedResults(Contact.class).into(contactConsumer),
  returningMappedResults(Account.class).into(accountConsumer),
  returningMappedResults(Invoice.class).into(invoiceConsumer))
```

These are just musings, but I hope they give you an idea of the motivation, and where this could go.